### PR TITLE
feat(one-on-one): allow oneOnOne input in startCheckIn mutation

### PR DESCRIPTION
**Pull request by igorlesnenko**
Fixes: #8561

**How to test:**

Enable org `oneOnOne` feature flag.

Edit this file to change variables

https://github.com/ParabolInc/parabol/blob/a3d5b02e6cea504d30136fa395b852d1d59744e0/packages/client/components/ActivityLibrary/ActivityDetailsSidebar.tsx#L153-L156

1) Change variables to set unknown email, start the `one on one` meeting

```
 StartCheckInMutation(
        atmosphere,
        {gcalInput, oneOnOneTeamInput: {
            email: 'igor+unknown@parabol.co',
            orgId: 'oFJQVmaSE8'
          }},
        {history, onError, onCompleted}
      )
```

- [ ] See that a new team was created and meeting was started
- [ ] See that invite was sent

2) Accept the invite, end the meeting. Keep the same code, start a new meeting again

- [ ] See that the team was reused, no new team was created

3) Test the same cases with an email of existing user

4) See that you still can create a meeting with `teamId`

### CODE REVIEW

File: packages/server/graphql/mutations/startCheckIn.ts [hunk 1]
```diff
@@ -17,35 +17,62 @@ import createGcalEvent from './helpers/createGcalEvent'
...
   async resolve(
     _source: unknown,
-    {teamId, gcalInput}: {teamId: string; gcalInput?: CreateGcalEventInputType},
-    {authToken, socketId: mutatorId, dataLoader}: GQLContext
+    {
+      teamId: existingTeamId,
+      gcalInput,
+      oneOnOneTeamInput
+    }: {
+      teamId: string
```
Comment (nickoferrall) [inline packages/server/graphql/mutations/startCheckIn.ts:49]: -1 `teamId` can now be undefined

File: packages/server/graphql/mutations/startCheckIn.ts [hunk 2]
```diff
@@ -17,35 +17,62 @@ import createGcalEvent from './helpers/createGcalEvent'
...
     const r = await getRethink()
+    const {authToken, socketId: mutatorId, dataLoader} = context
     const operationId = dataLoader.share()
     const subOptions = {mutatorId, operationId}
     // AUTH
     const viewerId = getUserId(authToken)
-    if (!isTeamMember(authToken, teamId)) {
-      return standardError(new Error('Team not found'), {userId: viewerId})
+
+    let teamId = existingTeamId
```
Comment (nickoferrall) [inline packages/server/graphql/mutations/startCheckIn.ts:62]: +1 I find the let a bit hard to follow. 

The first if check could check `existingTeamId` to help avoid `let`. 

Something like: 

```
    if (existingTeamId) {
      if (!isTeamMember(authToken, existingTeamId)) {
        return standardError(new Error('Team not found'), {userId: viewerId})
      }
      const unpaidError = await isStartMeetingLocked(existingTeamId, dataLoader)
      if (unpaidError) return standardError(new Error(unpaidError), {userId: viewerId})
    }
    const viewer = await dataLoader.get('users').loadNonNull(viewerId)
    const teamId = oneOnOneTeamInput ? await maybeCreateOneOnOneTeam(viewer, oneOnOneTeamInput, context) : existingTeamId
    if (!teamId) {
      return standardError(new Error('Must provide teamId or ad-hoc team'), {userId: viewerId})
    }
```

But that's just personal preference. Maybe you find that is harder to read.

File: packages/server/graphql/mutations/startCheckIn.ts [hunk 3]
```diff
@@ -17,35 +17,62 @@ import createGcalEvent from './helpers/createGcalEvent'
...
+      if (!isTeamMember(authToken, teamId)) {
+        return standardError(new Error('Team not found'), {userId: viewerId})
+      }
+      const unpaidError = await isStartMeetingLocked(teamId, dataLoader)
+      if (unpaidError) return standardError(new Error(unpaidError), {userId: viewerId})
+    } else if (oneOnOneTeamInput) {
+      const viewer = await dataLoader.get('users').loadNonNull(viewerId)
+      teamId = await maybeCreateOneOnOneTeam(viewer, oneOnOneTeamInput, context)
+    } else {
+      return standardError(new Error('Must provide teamId or ad-hoc team'), {userId: viewerId})
```
Comment (nickoferrall) [inline packages/server/graphql/mutations/startCheckIn.ts:74]: +1 nit but they would be providing `oneOnOneTeamInput` rather than an `ad-hoc team`

File: packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  await inviteToTeamHelper(invitees, teamId, undefined, context)
+
+  return teamId
+}
+
+const generateOneOnOneTeamName = (firstUserName: string, secondUserName: string) => {
+  return `${firstUserName} / ${secondUserName}`
+}
+
+const maybeCreteOneOnOneTeamFromEmail = async (
```
Comment (nickoferrall) [inline packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts:96]: +1 typo here: `Create` rather than `Crete`

File: packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+const createNewTeamAndInvite = async (
+  viewer: IUser,
+  name: string,
+  orgId: string,
+  invitees: string[],
+  context: GQLContext
+) => {
+  const {authToken} = context
+  const teamId = generateUID()
+  await createTeamAndLeader(viewer!, {id: teamId, isOnboardTeam: false, name, orgId})
```
Comment (nickoferrall) [inline packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts:80]: +1 can remove the `!`?

File: packages/server/graphql/public/types/CreateOneOnOneTeamInput.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import {GraphQLInputObjectType, GraphQLNonNull, GraphQLString} from 'graphql'
+
+const CreateOneOnOneTeamInput = new GraphQLInputObjectType({
+  name: 'CreateOneOnOneTeamInput',
+  fields: () => ({
+    userId: {
```
Comment (nickoferrall) [inline packages/server/graphql/public/types/CreateOneOnOneTeamInput.ts:6]: What's the use case where we'll provide a `userId` but not an `email`?

File: packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+const maybeCreteOneOnOneTeamFromEmail = async (
+  viewer: IUser,
+  email: string,
+  orgId: string,
+  context: GQLContext
+) => {
+  const teamName = generateOneOnOneTeamName(viewer.preferredName, email.split('@')[0]!)
+
+  const existingUsers = await getUsersByEmails([email])
```
Comment (nickoferrall) [inline packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts:104]: -1 let's use the `getUserByEmail` function that lives in the same file

File: packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+  return pg
+    .selectFrom('Team')
+    .selectAll()
+    .where('id', 'in', existingTeamIds)
+    .where('orgId', '=', orgId)
+    .where('isArchived', '=', false)
+    .execute()
+}
+
+const getExistingExactOneOnOneTeam = async (
```
Comment (nickoferrall) [inline packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts:62]: +1 why `exact` in the name rather than `getExistingOneOnOneTeam`?

File: packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+export default async function maybeCreateOneOnOneTeam(
+  viewer: IUser,
+  oneOnOneTeam: OneOnOneTeam,
+  context: GQLContext
+) {
+  const {email, userId, orgId} = oneOnOneTeam
+
+  if (email) {
+    return maybeCreteOneOnOneTeamFromEmail(viewer, email, orgId, context)
```
Comment (nickoferrall) [inline packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts:149]: -1 I feel like this file could be simplified by merging `maybeCreteOneOnOneTeamFromEmail` and `maybeCreateOneOnOneTeamFromUserId`.

It looks like the main difference is just getting the user by the userId or the email, which could be done as a ternary statement, like `const user = userId ? await getUserById(userId) : await getUsersByEmail(email!)`.

**Review by nickoferrall — CHANGES_REQUESTED** [review]
Tested and it's working for me!

Two notes:

1. Will we let users know in the UI that a new team will be created? They might be surprised that a Check-in creates a new team, but none of the other meetings do.
2. I need to refresh the page to see the newly created team at the moment

File: packages/server/graphql/public/types/CreateOneOnOneTeamInput.ts (hunk 6 above)
Comment (igorlesnenko) [reply packages/server/graphql/public/types/CreateOneOnOneTeamInput.ts:6]: I was thinking that when you select existing user in the user picker, we will provide a userId, however, I think I can simplify and just provide emails

File: packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts (hunk 9 above)
Comment (igorlesnenko) [reply packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts:149]: removed user option completely

**Comment by igorlesnenko** [comment]
> Will we let users know in the UI that a new team will be created? They might be surprised that a Check-in creates a new team, but none of the other meetings do.

We will show a hint, here is the separate issue for that: https://github.com/ParabolInc/parabol/issues/8562

> I need to refresh the page to see the newly created team at the moment 

Should work now

File: packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+  viewer: IUser,
+  oneOnOneTeam: OneOnOneTeam,
+  context: GQLContext
+) {
+  const {email, orgId} = oneOnOneTeam
+
+  const existingUser = await getUserByEmail(email)
+
+  if (!existingUser) {
+    const teamName = generateOneOnOneTeamName(viewer.preferredName, email.split('@')[0]!)
```
Comment (nickoferrall) [inline packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts:112]: +1 NIT: these two lines are the same as the 122-123 except for the second arg. Perhaps `generateOneOnOneTeamName` could handle the email split logic? It could take an object as an arg with `firstUserName: string`, `secondUserName?: string`, `secondUserEmail?: string`?

**Review by nickoferrall — APPROVED** [review]

File: packages/server/graphql/mutations/startCheckIn.ts [hunk 11]
```diff
@@ -17,35 +17,64 @@ import createGcalEvent from './helpers/createGcalEvent'
...
+
+    if (existingTeamId) {
+      if (!isTeamMember(authToken, existingTeamId)) {
+        return standardError(new Error('Team not found'), {userId: viewerId})
+      }
+      const unpaidError = await isStartMeetingLocked(existingTeamId, dataLoader)
+      if (unpaidError) return standardError(new Error(unpaidError), {userId: viewerId})
+    }
+    const viewer = await dataLoader.get('users').loadNonNull(viewerId)
+    const teamId = oneOnOneTeamInput
```
Comment (Dschoordsch) [inline packages/server/graphql/mutations/startCheckIn.ts:70]: -1 Please check that exactly one of `teamId`/`oneOnOneTeamInput` is set

File: packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    .without('reduction')
+    // Get all team members for each team
+    .eqJoin('group', r.table('TeamMember'), {index: 'teamId'})
+    .zip()
+    .group('teamId')
+    .pluck('userId')
+    .ungroup()
+    // Select only teams with exact given number of team members
+    .filter((row: RValue) => {
+      return row('reduction').count().eq(userIds.length)
```
Comment (Dschoordsch) [inline packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts:42]: -1 didn't we check this in l32 already?

File: packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+  orgId: string,
+  invitees: string[],
+  context: GQLContext
+) => {
+  const {authToken, socketId: mutatorId, dataLoader} = context
+  const {id: viewerId} = viewer
+  const operationId = dataLoader.share()
+  const subOptions = {mutatorId, operationId}
+  const teamId = generateUID()
+  await createTeamAndLeader(viewer, {id: teamId, isOnboardTeam: false, name, orgId})
```
Comment (Dschoordsch) [inline packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts:78]: +1 I think we should store the fact that it's a 1:1 team as well, so we can query and present it differently in the future.

**Review by Dschoordsch — CHANGES_REQUESTED** [review]
Not tested yet

File: packages/server/graphql/public/types/CreateOneOnOneTeamInput.ts [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+  name: 'CreateOneOnOneTeamInput',
+  fields: () => ({
+    email: {
+      type: new GraphQLNonNull(GraphQLString)
+    },
+    orgId: {
+      type: new GraphQLNonNull(GraphQLString)
+    }
+  })
+})
```
[the remark this answers is not in the document]
Comment (igorlesnenko) [reply packages/server/graphql/public/types/CreateOneOnOneTeamInput.ts:13]: I think I still need it to use in legacy `startCheckIn.ts`

File: packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts (hunk 12 above)
Comment (Dschoordsch) [reply packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts:42]: -1 Now that we store whether or not a team is a 1:1 team, we can use that here as well. We can just get a list of `teamId`s where all `userId`s are present and then just check in PG whether the team is a 1:1 team. This way we don't need to count at all. (These big rethink queries scare me)

**Review by Dschoordsch — CHANGES_REQUESTED** [review]

File: packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts (hunk 12 above)
Comment (Dschoordsch) [reply packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts:42]: For now I would ignore those 2 cases. If we see decent 1:1 usage, I would later restrict invites for those in the birthday cake version.
I'm open to discuss this.

Comment (igorlesnenko) [reply packages/server/graphql/mutations/helpers/maybeCreateOneOnOneTeam.ts]: @Dschoordsch I see nothing bad would happen if I ignore these cases for now, so I updated the PR.

**Review by Dschoordsch — APPROVED** [review]
